miniapps/blocksolvers/div_free_solver.cpp

```diff
@@ -0,0 +1,709 @@
+
+   int counter = 0;
+   for (int i = 0; i < tdof_agg.NumRows(); ++i)
+   {
+      bool agg_bdr = is_bdr[i] || is_shared.RowSize(i) || tdof_agg.RowSize(i)>1;
+      if (agg_bdr) { I[i+1] = I[i]; continue; }
+      I[i+1] = I[i] + 1;
+      J[counter++] = tdof_agg.GetRowColumns(i)[0];
+   }
+
+   double * D = new double[I[tdof_agg.NumRows()]];
+   std::fill_n(D, I[tdof_agg.NumRows()], 1.0);
+
+   SparseMatrix intdof_agg(I, J, D, tdof_agg.NumRows(), tdof_agg.NumCols());
+   return Transpose(intdof_agg);
+}
+
+void DFSDataCollector::MakeDofRelationTables(int level)
+{
+   Array<int> agg_starts(Array<int>(l2_0_fes_->GetDofOffsets(), 2));
+   auto& elem_agg = (const SparseMatrix&)*l2_0_fes_->GetUpdateOperator();
+   OperatorPtr agg_elem(Transpose(elem_agg));
+   SparseMatrix& agg_el = *agg_elem.As<SparseMatrix>();
+
+   el_l2dof_[level] = ElemToDof(*l2_fes_);
+   data_.agg_l2dof[level].Reset(Mult(agg_el, el_l2dof_[level]));
+
+   Array<int> bdr_tdofs;
+   hdiv_fes_->GetEssentialTrueDofs(all_bdr_attr_, bdr_tdofs);
+   auto tmp = AggToInteriorDof(bdr_tdofs, agg_el, ElemToDof(*hdiv_fes_),
+                               *hdiv_fes_->Dof_TrueDof_Matrix(), agg_starts);
+   data_.agg_hdivdof[level].Reset(tmp);
+}
+
+void DFSDataCollector::CollectData()
+{
+   --level_;
+
+   auto GetP = [this](OperatorPtr& P, unique_ptr<ParFiniteElementSpace>& cfes,
+                      ParFiniteElementSpace& fes, bool remove_zero)
+   {
+      fes.Update();
+      fes.GetTrueTransferOperator(*cfes, P);
+      if (remove_zero) { P.As<HypreParMatrix>()->Threshold(1e-16); }
+      this->level_ ? cfes->Update() : cfes.reset();
+   };
+
+   GetP(data_.P_hdiv[level_], coarse_hdiv_fes_, *hdiv_fes_, true);
+   GetP(data_.P_l2[level_], coarse_l2_fes_, *l2_fes_, false);
+   MakeDofRelationTables(level_);
+
+   GetP(data_.P_hcurl[level_], coarse_hcurl_fes_, *hcurl_fes_, true);
+
+   Vector trash1(hcurl_fes_->GetVSize()), trash2(hdiv_fes_->GetVSize());
+   ParDiscreteLinearOperator curl(hcurl_fes_.get(), hdiv_fes_.get());
+   curl.AddDomainInterpolator(new CurlInterpolator);
+   curl.Assemble();
+   curl.EliminateTrialDofs(ess_bdr_attr_, trash1, trash2);
+   curl.Finalize();
+   data_.C[level_].Reset(curl.ParallelAssemble());
+
+   if (level_ == 0) { DataFinalize(); }
+}
+
+void DFSDataCollector::DataFinalize()
+{
+   ParBilinearForm mass(l2_fes_.get());
+   mass.AddDomainIntegrator(new MassIntegrator());
+   mass.Assemble();
+   mass.Finalize();
+   OperatorPtr W(mass.ParallelAssemble());
+
+   for (int l = 0; l < data_.P_l2.Size(); ++l)
+   {
+      OperatorPtr PT_l2(data_.P_l2[l].As<HypreParMatrix>()->Transpose());
+      auto PTW = ParMult(PT_l2.As<HypreParMatrix>(), W.As<HypreParMatrix>(), true);
+      W.Reset(ParMult(PTW, data_.P_l2[l].As<HypreParMatrix>()));
+      auto cW_inv = new BlockDiagSolver(W, move(el_l2dof_[l+1]));
+      //TODO: maybe store the product is better if it needs to be applied in iterations
```

Comment:
I'd prefer not to have this TODO in released code. Can you decide which is actually better, choose it, and remove the comment?